File: backend/app/schemas/finance.py
from pydantic import BaseModel, Field, EmailStr
from typing import Optional, List, Dict, Any
from datetime import date, datetime


# ==========================================
# CURRENCY SCHEMAS
# ==========================================

class CurrencyBase(BaseModel):
    code: str = Field(..., max_length=10, description="ISO currency code e.g. USD, UGX, KES, EUR, GBP")
    name: str = Field(..., max_length=100, description="Currency name e.g. US Dollar, Uganda Shilling")
    symbol: str = Field(..., max_length=10, description="Currency symbol e.g. $, USh, KSh, €")
    exchange_rate_to_usd: float = Field(default=1.0, gt=0, description="Exchange rate relative to 1 USD")
    is_base_currency: bool = Field(default=False)
    is_active: bool = Field(default=True)


class CurrencyCreate(CurrencyBase):
    pass


class CurrencyUpdate(BaseModel):
    name: Optional[str] = None
    symbol: Optional[str] = None
    exchange_rate_to_usd: Optional[float] = Field(None, gt=0)
    is_base_currency: Optional[bool] = None
    is_active: Optional[bool] = None


class CurrencyResponse(CurrencyBase):
    id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# ==========================================
# COMPANY FINANCE SETTINGS SCHEMAS
# ==========================================

class CompanyFinanceSettingsBase(BaseModel):
    company_name: str = "ALLBOUND VACATIONS"
    legal_company_name: str = "Allbound Travel Services Limited"
    tagline: str = "Your Dream Holiday. Designed. Booked. Perfected."
    physical_address: Optional[str] = None
    phone: Optional[str] = None
    whatsapp: Optional[str] = None
    email: Optional[str] = None
    website: Optional[str] = None
    tin_number: Optional[str] = None
    company_registration_number: Optional[str] = None
    vat_number: Optional[str] = None
    bank_accounts: List[Dict[str, Any]] = Field(default_factory=list)
    mobile_money_accounts: List[Dict[str, Any]] = Field(default_factory=list)
    card_payment_info: Optional[str] = None
    default_invoice_notes: Optional[str] = None
    default_invoice_terms: Optional[str] = None
    default_receipt_notice: Optional[str] = None
    default_voucher_supplier_instructions: Optional[str] = None
    default_voucher_client_instructions: Optional[str] = None
    default_voucher_terms: Optional[str] = None


class CompanyFinanceSettingsUpdate(BaseModel):
    company_name: Optional[str] = None
    legal_company_name: Optional[str] = None
    tagline: Optional[str] = None
    physical_address: Optional[str] = None
    phone: Optional[str] = None
    whatsapp: Optional[str] = None
    email: Optional[str] = None
    website: Optional[str] = None
    tin_number: Optional[str] = None
    company_registration_number: Optional[str] = None
    vat_number: Optional[str] = None
    bank_accounts: Optional[List[Dict[str, Any]]] = None
    mobile_money_accounts: Optional[List[Dict[str, Any]]] = None
    card_payment_info: Optional[str] = None
    default_invoice_notes: Optional[str] = None
    default_invoice_terms: Optional[str] = None
    default_receipt_notice: Optional[str] = None
    default_voucher_supplier_instructions: Optional[str] = None
    default_voucher_client_instructions: Optional[str] = None
    default_voucher_terms: Optional[str] = None


class CompanyFinanceSettingsResponse(CompanyFinanceSettingsBase):
    id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# ==========================================
# INVOICE LINE ITEM SCHEMAS
# ==========================================

class InvoiceLineItemBase(BaseModel):
    category: str = Field(default="other", description="accommodation, transportation, activities, flights, other")
    title: str
    description: Optional[str] = None
    travel_date: Optional[date] = None
    quantity: float = 1.0
    unit_price: float = 0.0
    discount: float = 0.0
    tax_rate: float = 0.0
    tax_amount: float = 0.0
    total_amount: float = 0.0
    metadata_json: Optional[Dict[str, Any]] = Field(default_factory=dict)
    sort_order: int = 0


class InvoiceLineItemCreate(InvoiceLineItemBase):
    pass


class InvoiceLineItemResponse(InvoiceLineItemBase):
    id: int
    invoice_id: int

    class Config:
        from_attributes = True


# ==========================================
# INVOICE SCHEMAS
# ==========================================

class InvoiceBase(BaseModel):
    booking_id: Optional[int] = None
    quote_number: Optional[str] = None
    invoice_status: str = Field(default="draft", description="draft, issued, partially_paid, paid, overdue, cancelled")
    invoice_date: date
    booking_date: Optional[date] = None
    due_date: date
    currency: str = "USD"
    exchange_rate_to_usd: float = 1.0
    consultant_id: Optional[int] = None
    consultant_name: Optional[str] = None
    payment_terms: Optional[str] = None
    client_type: str = Field(default="individual", description="individual or corporate")
    client_details: Dict[str, Any] = Field(default_factory=dict)
    trip_summary: Dict[str, Any] = Field(default_factory=dict)
    subtotal: float = 0.0
    discount_amount: float = 0.0
    promotional_discount: float = 0.0
    taxable_amount: float = 0.0
    vat_amount: float = 0.0
    other_taxes_amount: float = 0.0
    service_fee: float = 0.0
    booking_fee: float = 0.0
    payment_processing_fee: float = 0.0
    total_amount: float = 0.0
    credit_applied: float = 0.0
    payment_methods_snapshot: Optional[Dict[str, Any]] = None
    notes: Optional[str] = None
    terms_and_conditions: Optional[str] = None


class InvoiceCreate(InvoiceBase):
    invoice_number: Optional[str] = None  # If None, auto-generated sequentially
    line_items: List[InvoiceLineItemCreate] = Field(default_factory=list)


class InvoiceFromBookingCreate(BaseModel):
    booking_id: int
    due_date: Optional[date] = None
    currency: Optional[str] = "USD"
    payment_terms: Optional[str] = None
    notes: Optional[str] = None


class InvoiceUpdate(BaseModel):
    invoice_status: Optional[str] = None
    quote_number: Optional[str] = None
    invoice_date: Optional[date] = None
    booking_date: Optional[date] = None
    due_date: Optional[date] = None
    currency: Optional[str] = None
    exchange_rate_to_usd: Optional[float] = None
    consultant_id: Optional[int] = None
    consultant_name: Optional[str] = None
    payment_terms: Optional[str] = None
    client_type: Optional[str] = None
    client_details: Optional[Dict[str, Any]] = None
    trip_summary: Optional[Dict[str, Any]] = None
    subtotal: Optional[float] = None
    discount_amount: Optional[float] = None
    promotional_discount: Optional[float] = None
    taxable_amount: Optional[float] = None
    vat_amount: Optional[float] = None
    other_taxes_amount: Optional[float] = None
    service_fee: Optional[float] = None
    booking_fee: Optional[float] = None
    payment_processing_fee: Optional[float] = None
    total_amount: Optional[float] = None
    credit_applied: Optional[float] = None
    payment_methods_snapshot: Optional[Dict[str, Any]] = None
    notes: Optional[str] = None
    terms_and_conditions: Optional[str] = None
    line_items: Optional[List[InvoiceLineItemCreate]] = None


# ==========================================
# PAYMENT RECEIPT SCHEMAS
# ==========================================

class PaymentReceiptBase(BaseModel):
    invoice_id: int
    booking_id: Optional[int] = None
    payment_reference: str
    receipt_date: date
    payment_date: datetime
    amount_received: float = Field(..., gt=0)
    currency: str = "USD"
    exchange_rate_to_usd: float = 1.0
    payment_method: str = Field(default="bank_transfer", description="bank_transfer, mobile_money, card, cash, eft, cheque, other")
    payment_provider: Optional[str] = None
    payment_status: str = Field(default="completed", description="completed, cleared, reversed")
    received_from: Dict[str, Any] = Field(default_factory=dict)
    notes: Optional[str] = None


class PaymentReceiptCreate(PaymentReceiptBase):
    receipt_number: Optional[str] = None  # If None, auto-generated sequentially
    amount_in_words: Optional[str] = None # If None, auto-calculated


class PaymentReceiptResponse(PaymentReceiptBase):
    id: int
    receipt_number: str
    amount_in_words: str
    payment_allocation: Dict[str, Any]
    verification_code: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class InvoiceResponse(InvoiceBase):
    id: int
    invoice_number: str
    amount_paid: float
    balance_due: float
    verification_token: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    line_items: List[InvoiceLineItemResponse] = Field(default_factory=list)
    receipts: List[PaymentReceiptResponse] = Field(default_factory=list)

    class Config:
        from_attributes = True


# ==========================================
# TRAVEL VOUCHER SCHEMAS
# ==========================================

class TravelVoucherBase(BaseModel):
    booking_id: Optional[int] = None
    invoice_id: Optional[int] = None
    confirmation_number: Optional[str] = None
    version: int = 1
    voucher_status: str = Field(default="confirmed", description="confirmed, pending, amended, cancelled")
    voucher_type: str = Field(default="general", description="accommodation, transportation, activity, safari, flight, general")
    issue_date: date
    issued_by_id: Optional[int] = None
    issued_by_name: Optional[str] = None
    supplier_details: Dict[str, Any] = Field(default_factory=dict)
    traveller_details: Dict[str, Any] = Field(default_factory=dict)
    service_details: Dict[str, Any] = Field(default_factory=dict)
    supplier_instructions: Optional[str] = None
    client_instructions: Optional[str] = None
    inclusions: List[str] = Field(default_factory=list)
    exclusions: List[str] = Field(default_factory=list)
    special_requests: List[Dict[str, Any]] = Field(default_factory=list)
    emergency_contacts: Optional[Dict[str, Any]] = None
    voucher_terms: Optional[str] = None


class TravelVoucherCreate(TravelVoucherBase):
    voucher_number: Optional[str] = None  # If None, auto-generated


class TravelVoucherFromBookingCreate(BaseModel):
    booking_id: int
    voucher_type: str = "general"
    supplier_details: Optional[Dict[str, Any]] = None
    issue_date: Optional[date] = None


class TravelVoucherUpdate(BaseModel):
    confirmation_number: Optional[str] = None
    voucher_status: Optional[str] = None
    voucher_type: Optional[str] = None
    issue_date: Optional[date] = None
    issued_by_name: Optional[str] = None
    supplier_details: Optional[Dict[str, Any]] = None
    traveller_details: Optional[Dict[str, Any]] = None
    service_details: Optional[Dict[str, Any]] = None
    supplier_instructions: Optional[str] = None
    client_instructions: Optional[str] = None
    inclusions: Optional[List[str]] = None
    exclusions: Optional[List[str]] = None
    special_requests: Optional[List[Dict[str, Any]]] = None
    emergency_contacts: Optional[Dict[str, Any]] = None
    voucher_terms: Optional[str] = None
    bump_version: bool = False  # If true, increments version number


class TravelVoucherResponse(TravelVoucherBase):
    id: int
    voucher_number: str
    verification_code: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# ==========================================
# PUBLIC VERIFICATION SCHEMAS
# ==========================================

class PublicVoucherVerificationResponse(BaseModel):
    is_valid: bool
    voucher_number: str
    booking_number: Optional[str] = None
    version: int
    voucher_status: str
    voucher_type: str
    issue_date: date
    supplier_name: str
    lead_traveller: str
    passenger_count: int
    service_summary: str
    service_dates: str
    verification_timestamp: datetime


class PublicReceiptVerificationResponse(BaseModel):
    is_valid: bool
    receipt_number: str
    invoice_number: Optional[str] = None
    booking_number: Optional[str] = None
    receipt_date: date
    amount_received: float
    currency: str
    payment_method: str
    payment_status: str
    payer_name: str
    verification_timestamp: datetime


# ==========================================
# DASHBOARD STATS & ACTIONS
# ==========================================

class FinanceDashboardStats(BaseModel):
    total_invoiced_usd: float
    total_collected_usd: float
    total_outstanding_usd: float
    overdue_invoices_count: int
    pending_vouchers_count: int
    invoices_by_status: Dict[str, int]
    recent_receipts: List[PaymentReceiptResponse]
    recent_invoices: List[InvoiceResponse]


class SendEmailRequest(BaseModel):
    recipient_email: EmailStr
    recipient_name: Optional[str] = None
    subject: Optional[str] = None
    message: Optional[str] = None
    copy_me: bool = True

File: frontend/src/lib/types/finance.ts
export interface Currency {
  id: number;
  code: string;
  name: string;
  symbol: string;
  exchange_rate_to_usd: number;
  is_base_currency: boolean;
  is_active: boolean;
  created_at?: string;
  updated_at?: string;
}

export interface BankAccount {
  bank_name: string;
  account_name: string;
  account_number: string;
  branch?: string;
  swift_bic?: string;
  currency: string;
  bank_address?: string;
  is_primary?: boolean;
}

export interface MobileMoneyAccount {
  network: string;
  merchant_number: string;
  account_name: string;
  currency?: string;
  is_primary?: boolean;
}

export interface CompanyFinanceSettings {
  id: number;
  company_name: string;
  legal_company_name: string;
  tagline: string;
  physical_address?: string;
  phone?: string;
  whatsapp?: string;
  email?: string;
  website?: string;
  tin_number?: string;
  company_registration_number?: string;
  vat_number?: string;
  bank_accounts: BankAccount[];
  mobile_money_accounts: MobileMoneyAccount[];
  card_payment_info?: string;
  default_invoice_notes?: string;
  default_invoice_terms?: string;
  default_receipt_notice?: string;
  default_voucher_supplier_instructions?: string;
  default_voucher_client_instructions?: string;
  default_voucher_terms?: string;
}

export interface ClientDetails {
  full_name?: string;
  name?: string;
  email?: string;
  telephone?: string;
  phone?: string;
  address?: string;
  country?: string;
  company_name?: string;
  company?: string;
  contact_person?: string;
  tin_vat?: string;
  purchase_order_number?: string;
  [key: string]: any;
}

export interface TripSummary {
  lead_traveller?: string;
  destinations?: string;
  travel_type?: string;
  travel_start_date?: string | null;
  travel_end_date?: string | null;
  adults?: number;
  children?: number;
  infants?: number;
  duration_days?: number | null;
  duration_nights?: number | null;
  accommodation_category?: string;
  meal_plan?: string;
  transportation_type?: string;
  tour_package_name?: string;
}

export interface InvoiceLineItem {
  id?: number;
  invoice_id?: number;
  category: 'accommodation' | 'transportation' | 'activities' | 'flights' | 'other' | string;
  title: string;
  description?: string;
  travel_date?: string | null;
  quantity: number;
  unit_price: number;
  discount: number;
  tax_rate: number;
  tax_amount: number;
  total_amount: number;
  cost_price?: number;
  supplier_id?: number | null;
  metadata_json?: Record<string, any>;
  sort_order?: number;
}

export interface Invoice {
  id: number;
  invoice_number: string;
  booking_id?: number | null;
  quote_number?: string;
  invoice_status: 'draft' | 'issued' | 'partially_paid' | 'paid' | 'overdue' | 'cancelled';
  invoice_date: string;
  booking_date?: string | null;
  due_date: string;
  currency: string;
  exchange_rate_to_usd: number;
  consultant_id?: number | null;
  consultant_name?: string;
  payment_terms?: string;
  client_type: 'individual' | 'corporate';
  client_details: ClientDetails;
  trip_summary: TripSummary;
  subtotal: number;
  discount_amount: number;
  promotional_discount: number;
  taxable_amount: number;
  vat_amount: number;
  other_taxes_amount: number;
  service_fee: number;
  booking_fee: number;
  payment_processing_fee: number;
  total_amount: number;
  amount_paid: number;
  credit_applied: number;
  balance_due: number;
  payment_methods_snapshot?: {
    bank_accounts?: BankAccount[];
    mobile_money_accounts?: MobileMoneyAccount[];
    card_payment_info?: string;
  };
  notes?: string;
  terms_and_conditions?: string;
  verification_token: string;
  created_at?: string;
  updated_at?: string;
  line_items: InvoiceLineItem[];
  receipts: PaymentReceipt[];
}

export interface InvoiceCreateRequest {
  invoice_number?: string;
  booking_id?: number | null;
  quote_number?: string;
  invoice_status?: string;
  invoice_date: string;
  booking_date?: string | null;
  due_date: string;
  currency: string;
  exchange_rate_to_usd?: number;
  consultant_id?: number | null;
  consultant_name?: string;
  payment_terms?: string;
  client_type: 'individual' | 'corporate';
  client_details: ClientDetails;
  trip_summary: TripSummary;
  subtotal: number;
  discount_amount: number;
  promotional_discount: number;
  taxable_amount: number;
  vat_amount: number;
  other_taxes_amount: number;
  service_fee: number;
  booking_fee: number;
  payment_processing_fee: number;
  total_amount: number;
  credit_applied: number;
  payment_methods_snapshot?: any;
  notes?: string;
  terms_and_conditions?: string;
  line_items: InvoiceLineItem[];
}

export type InvoiceUpdateRequest = Partial<InvoiceCreateRequest>;

export interface PaymentAllocation {
  description?: string;
  invoice_amount?: number;
  previously_paid?: number;
  this_payment?: number;
  balance?: number;
  total_paid?: number;
}

export interface PaymentReceipt {
  id: number;
  receipt_number: string;
  invoice_id: number;
  booking_id?: number | null;
  payment_reference: string;
  receipt_date: string;
  payment_date: string;
  amount_received: number;
  amount_in_words: string;
  currency: string;
  exchange_rate_to_usd: number;
  payment_method: 'bank_transfer' | 'mobile_money' | 'card' | 'cash' | 'eft' | 'cheque' | 'other' | string;
  payment_provider?: string;
  payment_status: 'completed' | 'cleared' | 'reversed' | string;
  received_from: ClientDetails;
  payment_allocation: PaymentAllocation;
  verification_code: string;
  notes?: string;
  created_at?: string;
  updated_at?: string;
}

export interface PaymentReceiptCreateRequest {
  receipt_number?: string;
  invoice_id: number;
  booking_id?: number | null;
  payment_reference: string;
  receipt_date: string;
  payment_date: string;
  amount_received: number;
  amount_in_words?: string;
  currency?: string;
  exchange_rate_to_usd?: number;
  payment_method: string;
  payment_provider?: string;
  payment_status?: string;
  received_from?: ClientDetails;
  notes?: string;
}

export interface SupplierDetails {
  supplier_name: string;
  company_name?: string;
  hotel_name?: string;
  contact_person?: string;
  reservation_email?: string;
  telephone?: string;
  address?: string;
  supplier_confirmation_number?: string;
}

export interface TravellerDetails {
  lead_traveller: string;
  travellers_list: string[];
  adults: number;
  children: number;
  children_ages?: (number | null)[];
  infants?: number;
  nationality?: string;
  passport_info?: string;
  country_of_origin?: string;
}

export interface SpecialRequestItem {
  request: string;
  status: 'confirmed' | 'subject_to_availability' | 'requested';
}

export interface TravelVoucher {
  id: number;
  voucher_number: string;
  booking_id?: number | null;
  invoice_id?: number | null;
  confirmation_number?: string;
  version: number;
  voucher_status: 'confirmed' | 'pending' | 'amended' | 'cancelled' | string;
  voucher_type: 'accommodation' | 'transportation' | 'activity' | 'safari' | 'flight' | 'general' | string;
  issue_date: string;
  issued_by_id?: number | null;
  issued_by_name?: string;
  supplier_details: SupplierDetails;
  traveller_details: TravellerDetails;
  service_details: Record<string, any>;
  supplier_instructions?: string;
  client_instructions?: string;
  inclusions: string[];
  exclusions: string[];
  special_requests: SpecialRequestItem[];
  emergency_contacts?: Record<string, string>;
  voucher_terms?: string;
  verification_code: string;
  created_at?: string;
  updated_at?: string;
}

export interface TravelVoucherCreateRequest {
  voucher_number?: string;
  booking_id?: number | null;
  invoice_id?: number | null;
  confirmation_number?: string;
  version?: number;
  voucher_status?: string;
  voucher_type: string;
  issue_date: string;
  supplier_details: SupplierDetails;
  traveller_details: TravellerDetails;
  service_details: Record<string, any>;
  supplier_instructions?: string;
  client_instructions?: string;
  inclusions?: string[];
  exclusions?: string[];
  special_requests?: SpecialRequestItem[];
  emergency_contacts?: Record<string, string>;
  voucher_terms?: string;
}

export interface TravelVoucherUpdateRequest extends Partial<TravelVoucherCreateRequest> {
  bump_version?: boolean;
}

export interface FinanceDashboardStats {
  total_invoiced_usd: number;
  total_collected_usd: number;
  total_outstanding_usd: number;
  overdue_invoices_count: number;
  pending_vouchers_count: number;
  invoices_by_status: Record<string, number>;
  recent_receipts: PaymentReceipt[];
  recent_invoices: Invoice[];
}

export interface PublicVoucherVerificationResponse {
  is_valid: boolean;
  voucher_number: string;
  booking_number?: string;
  version: number;
  voucher_status: string;
  voucher_type: string;
  issue_date: string;
  supplier_name: string;
  lead_traveller: string;
  passenger_count: number;
  service_summary: string;
  service_dates: string;
  verification_timestamp: string;
}

export interface PublicReceiptVerificationResponse {
  is_valid: boolean;
  receipt_number: string;
  invoice_number?: string;
  booking_number?: string;
  receipt_date: string;
  amount_received: number;
  currency: string;
  payment_method: string;
  payment_status: string;
  payer_name: string;
  verification_timestamp: string;
}

// ==========================================
// CLIENT CRM TYPES
// ==========================================

export interface Client {
  id: number;
  client_type: 'individual' | 'corporate';
  first_name?: string;
  last_name?: string;
  company_name?: string;
  contact_person?: string;
  email: string;
  phone?: string;
  alt_phone?: string;
  country_of_origin?: string;
  address?: string;
  city?: string;
  postal_code?: string;
  tin_number?: string;
  vat_number?: string;
  passport_number?: string;
  nationality?: string;
  dietary_requirements?: string;
  special_notes?: string;
  is_active: boolean;
  display_name: string;
  total_invoiced_usd: number;
  total_paid_usd: number;
  outstanding_balance_usd: number;
  invoices_count: number;
  bookings_count: number;
  created_at?: string;
  updated_at?: string;
}

export interface StatementTransaction {
  date: string;
  type: 'INVOICE' | 'PAYMENT';
  reference_number: string;
  description: string;
  currency: string;
  debit: number;
  credit: number;
  running_balance: number;
}

export interface ClientStatement {
  client: Client;
  statement_date: string;
  start_date?: string;
  end_date?: string;
  opening_balance: number;
  total_billed: number;
  total_paid: number;
  closing_balance: number;
  transactions: StatementTransaction[];
}

// ==========================================
// SUPPLIERS & PAYABLES TYPES
// ==========================================

export interface Supplier {
  id: number;
  name: string;
  supplier_code: string;
  category: 'lodge_hotel' | 'safari_operator' | 'transporter' | 'airline' | 'park_authority' | 'guide' | 'other' | string;
  contact_person?: string;
  email?: string;
  phone?: string;
  whatsapp?: string;
  physical_address?: string;
  country?: string;
  currency: string;
  tax_pin_number?: string;
  bank_details?: Record<string, any>;
  mobile_money_details?: Record<string, any>;
  payment_terms?: string;
  rating?: number;
  notes?: string;
  is_active: boolean;
  total_billed_usd: number;
  total_paid_usd: number;
  balance_payable_usd: number;
  pending_bills_count: number;
  created_at?: string;
  updated_at?: string;
}

export interface SupplierPayment {
  id: number;
  payment_number: string;
  supplier_bill_id: number;
  supplier_id: number;
  payment_date: string;
  amount_paid: number;
  currency: string;
  exchange_rate_to_usd: number;
  payment_method: string;
  reference_code: string;
  disbursed_from_account?: string;
  notes?: string;
  created_at?: string;
}

export interface SupplierBill {
  id: number;
  bill_number: string;
  supplier_reference?: string;
  supplier_id: number;
  supplier_name?: string;
  booking_id?: number;
  invoice_id?: number;
  bill_date: string;
  due_date: string;
  service_date?: string;
  currency: string;
  exchange_rate_to_usd: number;
  amount_billed: number;
  amount_paid: number;
  balance_payable: number;
  status: 'pending' | 'partially_paid' | 'paid' | 'overdue' | 'cancelled';
  category: string;
  description?: string;
  notes?: string;
  attachment_url?: string;
  created_at?: string;
  updated_at?: string;
  payments: SupplierPayment[];
}

export interface SupplierLedgerTransaction {
  date: string;
  type: 'BILL' | 'PAYMENT';
  reference_number: string;
  supplier_reference?: string;
  description: string;
  currency: string;
  bill_amount: number;
  paid_amount: number;
  running_payable: number;
}

export interface SupplierLedger {
  supplier: Supplier;
  statement_date: string;
  start_date?: string;
  end_date?: string;
  total_billed: number;
  total_paid: number;
  closing_payable: number;
  transactions: SupplierLedgerTransaction[];
}

// ==========================================
// FINANCIAL REPORTS TYPES
// ==========================================

export interface SalesReport {
  start_date?: string;
  end_date?: string;
  total_invoiced_usd: number;
  total_collected_usd: number;
  total_outstanding_usd: number;
  invoices_count: number;
  average_order_value_usd: number;
  period_breakdown: {
    period: string;
    invoices_count: number;
    gross_revenue_usd: number;
    discount_usd: number;
    net_revenue_usd: number;
    collected_usd: number;
    outstanding_usd: number;
  }[];
  destination_breakdown: {
    destination: string;
    bookings_count: number;
    total_sales_usd: number;
    percentage_of_total: number;
  }[];
  consultant_breakdown: {
    consultant_name: string;
    invoices_count: number;
    total_sales_usd: number;
  }[];
}

export interface AgingBucket {
  bucket_label: string;
  count: number;
  total_amount_usd: number;
  percentage: number;
}

export interface ReceivablesAgingReport {
  as_of_date: string;
  total_receivable_usd: number;
  buckets: AgingBucket[];
  overdue_invoices: {
    invoice_id: number;
    invoice_number: string;
    client_name: string;
    invoice_date: string;
    due_date: string;
    days_overdue: number;
    currency: string;
    total_amount: number;
    amount_paid: number;
    balance_due: number;
    balance_due_usd: number;
    status: string;
  }[];
}

export interface PayablesAgingReport {
  as_of_date: string;
  total_payable_usd: number;
  buckets: AgingBucket[];
  pending_bills: {
    bill_id: number;
    bill_number: string;
    supplier_name: string;
    bill_date: string;
    due_date: string;
    days_overdue: number;
    currency: string;
    amount_billed: number;
    amount_paid: number;
    balance_payable: number;
    balance_payable_usd: number;
    status: string;
  }[];
}

export interface ProfitabilityItem {
  booking_id?: number;
  invoice_id: number;
  invoice_number: string;
  client_name: string;
  service_description: string;
  destination?: string;
  revenue_usd: number;
  cost_usd: number;
  gross_profit_usd: number;
  gross_margin_percent: number;
}

export interface ProfitabilityReport {
  start_date?: string;
  end_date?: string;
  total_revenue_usd: number;
  total_cost_usd: number;
  total_gross_profit_usd: number;
  average_margin_percent: number;
  items: ProfitabilityItem[];
}

export interface CashFlowReport {
  start_date?: string;
  end_date?: string;
  total_inflow_usd: number;
  total_outflow_usd: number;
  net_cash_flow_usd: number;
  inflows_by_method: Record<string, number>;
  outflows_by_method: Record<string, number>;
  daily_timeline: {
    date: string;
    inflow_usd: number;
    outflow_usd: number;
    net_usd: number;
  }[];
}

export interface InvoiceProfitability {
  invoice_id: number;
  invoice_number: string;
  currency: string;
  exchange_rate_to_usd: number;
  total_revenue: number;
  total_revenue_usd: number;
  total_expenses: number;
  total_expenses_usd: number;
  gross_profit: number;
  gross_profit_usd: number;
  gross_margin_percent: number;
  bills_count: number;
  supplier_bills: SupplierBill[];
}

